README.md
# SafePreviewDevice

<p align="center">
    <a href="#">
        <img src="https://img.shields.io/badge/swift-5.1-orange.svg" alt="Swift: 5.1">
    </a>
    <a href="https://github.com/posix88/SafePreviewDevice/releases">
        <img src="https://img.shields.io/badge/version-0.1.0-blue.svg"
        alt="Version: 0.1.0">
    </a>
    <a href="#">
    <img src="https://img.shields.io/badge/Platforms-iOS%20|%20tvOS%20|%20watchOS|%20macOS-green.svg"
        alt="Platforms: iOS – tvOS – watchOS - macOS">
    </a>
    <a href="https://github.com/piknotech/SFSafeSymbols/blob/stable/LICENSE.md">
        <img src="https://img.shields.io/badge/license-MIT-lightgrey.svg" alt="License: MIT">
    </a>
    <br />
    <a href="https://github.com/apple/swift-package-manager">
        <img src="https://img.shields.io/badge/SwiftPM-compatible-brightgreen.svg" alt="SwiftPM: Compatible">
    </a>
    <a href="https://cocoapods.org/pods/SafePreviewDevice">
    <img src="https://img.shields.io/badge/CocoaPods-compatible-4BC51D.svg?style=flat" alt="CocoaPods: Compatible">
    </a>
</p>

## Motivation

At WWDC 2019, Apple announced SwiftUI a new library for building UI in a simple and fast way.  
Xcode’s SwiftUI preview lets us show our application design in multiple screen sizes at the same time using the `.previewDevice()` modifier. Currently, this needs to be provided with the exact name of the device you want to test:  e.g. “iPhone XS Max”.

```swift
.previewDevice(PreviewDevice(rawValue: "iPhone XS Max"))
```
or even better

```swift
.previewDevice(.iPhone7)
```

This framework comes to rescue for everyone that are looking for an easy way to select their preview device or they just doesn't remember every iOS 13 device name (like me 😀)

## System Requirements

Supported platforms are `iOS (13.0)`, `tvOS (13.0)`, `watchOS (6.0)` and `macOS (10.15)` and of course `Xcode 11`.

## Installation

`SafePreviewDevice` can be installed via Swift Package Manager.

### Swift Package Manager

To integrate using Apple's Swift package manager, add the following as a dependency to your `Package.swift`:

```swift
.package(url: "https://github.com/posix88/SafePreviewDevice.git", .upToNextMajor(from: "0.1.0"))
```

and  run `swift package update`.

Otherwise you can simply go to File -> Swift Packages -> Add Package Dependency and paste the repo's url: `https://github.com/posix88/SafePreviewDevice`

## Usage

Simply import this framework in your `SwiftUI` file and add one of this `View` modifier

### Change the preview device
```swift
.previewDevice(.iPhone7)
```
This modifier will change the device in Xcode SwiftUI preview.

### Change the preview device name
```swift
.previewDeviceName(.iPhone7)
```
This modifier will change the device name in Xcode SwiftUI preview.

## Contributing

- If you **need help** or you'd like to **ask a general question**, open an issue.
- If you **found a bug**, open an issue.
- If you **have a feature request**, open an issue.
- If you **want to contribute**, submit a pull request.


## Acknowledgements

Made with ❤️ in Milan by [Antonino Musolino](https://twitter.com/NinoMusolino).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.1.0] Hello Release 👋

### Added

- Two `View` extensions:
    `previewDevice(_ device: Devices)`
    `previewDeviceName(_ device: Devices)`
    that will semplify your PreviewDevice choice.

### Changed

- None

### Fixed

- None


Tests/LinuxMain.swift
import XCTest

import SafePreviewDeviceTests

var tests = [XCTestCaseEntry]()
tests += SafePreviewDeviceTests.allTests()
XCTMain(tests)


Tests/SafePreviewDeviceTests/XCTestManifests.swift
import XCTest

#if !canImport(ObjectiveC)
public func allTests() -> [XCTestCaseEntry] {
    return [
        testCase(SafePreviewDeviceTests.allTests),
    ]
}
#endif


Tests/SafePreviewDeviceTests/SafePreviewDeviceTests.swift
import XCTest
@testable import SafePreviewDevice

#if !os(watchOS) && !os(macOS)

import XCTest

#if canImport(SwiftUI)

import SwiftUI

class SafePreviewTests: XCTestCase {
    
    func testInit() {
        if #available(iOS 13.0, OSX 10.15, tvOS 13.0, watchOS 6.0, *) {
            Devices.allCases.forEach { device in
                print("Testing existence of \(device.rawValue)")
                XCTAssertNotNil(PreviewDevice(device: device))
            }
        } else {
            XCTFail("Latest version of Apple OS are needed to test this package")
        }
    }
}

#else

class JustFail: XCTestCase {
    func justFail() {
        XCTFail("SwiftUI must be available when testing.")
    }
}

#endif

#endif


Package.swift
// swift-tools-version:5.2
// The swift-tools-version declares the minimum version of Swift required to build this package.

import PackageDescription

let package = Package(
    name: "SafePreviewDevice",
    platforms: [.iOS(.v13), .tvOS(.v13), .watchOS(.v6), .macOS(.v10_15)],
    products: [
        // Products define the executables and libraries produced by a package, and make them visible to other packages.
        .library(
            name: "SafePreviewDevice",
            targets: ["SafePreviewDevice"]),
    ],
    dependencies: [
        // Dependencies declare other packages that this package depends on.
        // .package(url: /* package url */, from: "1.0.0"),
    ],
    targets: [
        // Targets are the basic building blocks of a package. A target can define a module or a test suite.
        // Targets can depend on other targets in this package, and on products in packages which this package depends on.
        .target(
            name: "SafePreviewDevice",
            dependencies: []),
        .testTarget(
            name: "SafePreviewDeviceTests",
            dependencies: ["SafePreviewDevice"]),
    ]
)


Sources/SafePreviewDevice/ViewExtension.swift
//
//  ViewExtension.swift
//  SafePreviewDevice
//
//  Created by MUSOLINO Antonino on 29/04/2020.
//

#if canImport(SwiftUI)
import SwiftUI

@available(iOS 13.0, OSX 10.15, tvOS 13.0, watchOS 6.0, *)
public extension View {
    
    /// Overrides the device for a preview.
    /// - Parameter device: the choosen device
    /// - Returns: The Xcode preview canvas `View`
    func previewDevice(_ device: Devices) -> some View {
        let device: PreviewDevice? = PreviewDevice(device: device)
        return previewDevice(device)
    }
    
    /// Provides a user visible name shown in the editor.
    /// - Parameter device: the choosen device
    /// - Returns: The visible name label below preview canvas
    func previewDeviceName(_ device: Devices) -> some View {
        previewDisplayName(device.rawValue)
    }
}

#endif


Sources/SafePreviewDevice/Devices.swift
//
//  Devices.swift
//  SafePreviewDevice
//
//  Created by MUSOLINO Antonino on 29/04/2020.
//
#if canImport(SwiftUI)

@available(iOS 13.0, macOS 10.15, tvOS 13.0, watchOS 6.0, *)
public enum Devices: String, CaseIterable {
    case mac = "Mac"
    case iPhone7 = "iPhone 7"
    case iPhone7Plus = "iPhone 7 Plus"
    case iPhone8 = "iPhone 8"
    case iPhone8Plus = "iPhone 8 Plus"
    case iPhoneSE = "iPhone SE"
    case iPhoneX = "iPhone X"
    case iPhoneXs = "iPhone Xs"
    case iPhoneXsMax = "iPhone Xs Max"
    case iPhoneXr = "iPhone Xʀ"
    case iPadMini4 = "iPad mini 4"
    case iPadAir2 = "iPad Air 2"
    case iPadPro9Inch = "iPad Pro (9.7-inch)"
    case iPadPro12InchV1 = "iPad Pro (12.9-inch)"
    case iPadV5 = "iPad (5th generation)"
    case iPadPro12InchV2 = "iPad Pro (12.9-inch) (2nd generation)"
    case iPadPro10Inch = "iPad Pro (10.5-inch)"
    case iPadV6 = "iPad (6th generation)"
    case iPadPro11Inch = "iPad Pro (11-inch)"
    case iPadPro12InchV3 = "iPad Pro (12.9-inch) (3rd generation)"
    case iPadMini5 = "iPad mini (5th generation)"
    case iPadAirV3 = "iPad Air (3rd generation)"
    case appleTV = "Apple TV"
    case appleTV4K = "Apple TV 4K"
    case appleTV4Kdownscaled = "Apple TV 4K (at 1080p)"
    case watch2Small = "Apple Watch Series 2 - 38mm"
    case watch2Large = "Apple Watch Series 2 - 42mm"
    case watch3Small = "Apple Watch Series 3 - 38mm"
    case watch3Large = "Apple Watch Series 3 - 42mm"
    case watch4Small = "Apple Watch Series 4 - 40mm"
    case watch4Large = "Apple Watch Series 4 - 44mm"
}

#endif
